scalafix-core/src/main/scala/scalafix/patch/Patch.scala

```diff
@@ -0,0 +1,186 @@
+package scalafix
+package patch
+
+import metaconfig._
+import scala.collection.immutable.Seq
+import scala.meta._
+import scala.meta.contrib._
+import scala.meta.internal.ast.Helpers._
+import scala.meta.tokens.Token
+import scala.meta.tokens.Token
+import scalafix.config._
+import scalafix.syntax._
+import scalafix.patch.TokenPatch.Add
+import scalafix.patch.TokenPatch.Remove
+import scalafix.patch.TreePatch.RenamePatch
+import scalafix.patch.TreePatch.Replace
+
+import difflib.DiffUtils
+import org.scalameta.logger
+
+sealed abstract class Patch {
+  // NOTE: potential bottle-neck, this might be very slow for large
+  // patches. We might want to group related patches and enforce some ordering.
+  def +(other: Patch): Patch =
+    if (this eq other) this
+    else {
+      (this, other) match {
+        case (InCtx(ap, ac, am), b @ InCtx(bp, bc, bm)) =>
+          if (ac ne bc) ???
+          else InCtx(ap + bp, ac, am.orElse(bm))
```

Comment:
Good catch, I now fail-fast with a detailed error message if `am ne bm`. This is an invariant that should never happen.